src/mlpack/core/data/load_image.hpp

```diff
@@ -0,0 +1,150 @@
+#define MLPACK_CORE_DATA_LOAD_IMAGE_HPP
+
+#include <string>
+#include <iostream>
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/log.hpp>
+#include <mlpack/prereqs.hpp>
+
+#ifdef HAS_FILESYSTEM
+#include <experimental/filesystem>
+#endif
+
+#include "extension.hpp"
+
+#define STB_IMAGE_IMPLEMENTATION
+#include <stb_image.h>
+
+namespace mlpack {
+namespace data {
+
+/**
+ * Loads a matrix with image. It also supports loading image from 
+ * an entire directory.
+ * @code
+ * arma::Mat<unsigned char> img;
+ * data::LoadImage loader;
+ * loader.Load("test_image.png", std::move(img));
+ * @endcode
+ *
+ * The supported types of files are:
+ * 
+ * - JPEG baseline & progressive (12 bpc/arithmetic not supported,
+ *    same as stock IJG lib)
+ * - PNG 1/2/4/8/16-bit-per-channel
+ * - TGA (not sure what subset, if a subset)
+ * - BMP non-1bpp, non-RLE
+ * - PSD (composited view only, no extra channels, 8/16 bit-per-channel)
+ * - GIF (*comp always reports as 4-channel)
+ * - HDR (radiance rgbE format)
+ * - PIC (Softimage PIC)
+ * - PNM (PPM and PGM binary only).
+ */
+class LoadImage
+{
+ public:
+  /**
+   * LoadImage default constructor.
+   */
+  LoadImage();
+
+  /**
+   * Instantiate the LoadImage object with the image width, height, channels..
+   *
+   * @param width Matrix width for the output matrix.
+   * @param height Matrix height for the output matrix.
+   * @param channels Matrix channels for the output matrix.
+   */
+  LoadImage(int width, int height, int channels);
+
+  /**
+   * Checks if the given image filename is supported.
+   *
+   * @param filename Name of the image file.
+   * @return Boolean value indicating success if it is an image.
+   */
+  bool IsImageFile(std::string& fileName);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param fileName Name of the image file.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(std::string& fileName, arma::Mat<unsigned char>&& outputMatrix);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param fileName Name of the image file.
+   * @param width Width of the image file.
+   * @param height Height of the image file.
+   * @param channels Channels of the image file.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(std::string& fileName,
+          arma::Mat<unsigned char>&& outputMatrix,
+          int *width,
+          int *height,
+          int *channels);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param files A vector containing names of the image file to be loaded.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(std::vector<std::string>& files,
+    arma::Mat<unsigned char>&& outputMatrix);
+
+  /**
+   * Load the image file into the given matrix.
```

Comment:
It would probably be good to point out that this fails if `HAS_FILESYSTEM` is not available (I guess, that just means C++17 support is not available?).  And it may be better to throw a `std::runtime_error` instead of returning a bool.

Reply:
Looks with gcc you also have to add ``-lstdc++fs`` when you want filesystem. But since we have this nice build bots we should be able to cover a bunch of different setups.

Reply:
Looks like we could something similar as done here: https://github.com/gulrak/filesystem to check for support:

```c++
#if defined(__cplusplus) && __cplusplus >= 201703L && defined(__has_include) && __has_include(<filesystem>)
#include <filesystem>
namespace fs = std::filesystem;
#else
#include <ghc/filesystem.hpp>
namespace fs = ghc::filesystem;
#endif
```